packages/@uppy/xhr-upload/src/index.js

```diff
@@ -99,21 +99,85 @@ export default class XHRUpload extends BasePlugin {
       /**
        * Check if the response from the upload endpoint indicates that the upload was successful.
        *
-       * @param {number} status the response status code
+       * @param {number} status
+       * @param {string} responseText
        */
-      validateStatus (status) {
+      // eslint-disable-next-line no-unused-vars
+      validateStatus (status, responseText) {
         return status >= 200 && status < 300
       },
     }
 
     this.opts = { ...defaultOptions, ...opts }
     this.i18nInit()
 
-    // Simultaneous upload limiting is shared across all uploads with this plugin.
-    if (internalRateLimitedQueue in this.opts) {
-      this.requests = this.opts[internalRateLimitedQueue]
-    } else {
-      this.requests = new RateLimitedQueue(this.opts.limit)
+    this.uppy.queue.concurrency = this.opts.limit
+
+    /**
+     * xhr-upload wrapper for `fetcher` to handle user options
+     * `validateStatus`, `getResponseError`, `getResponseData`
+     * and to emit `upload-progress`, `upload-error`, and `upload-success` events.
+     *
+     * @param {import('@uppy/core').UppyFile[]} files
+     */
+    this.#uppyFetch = (files) => {
+      /** @type {typeof fetcher} */
+      return async (url, options) => {
+        try {
+          const response = await fetcher(url, {
+            ...options,
+            onTimeout: () => {
+              const seconds = Math.ceil(options.timeout / 1000)
+              const error = new Error(this.i18n('uploadStalled', { seconds }))
```

Comment:
Not sure if we should put i18n strings in error messages. Error objects are not user-facing but developer-facing. Having error messages in different languages could make debugging production issues harder.

Reply:
This was already the case, I just moved the code. It is user facing though, people view this message.

Reply:
Ok, I see. I just don't think that Error objects (and their `message`) should be user facing, but yea we can fix it in a diferent pr